bad/source/lib.rs

```diff
@@ -11,31 +17,47 @@ pub enum ProgramSource {
 }
 
 pub struct CompilationConfiguration {
-	pub sources: Vec<ProgramSource>,
+	
 }
 
 #[derive(Debug)]
-pub enum Token {
-	LeftParen,
-	RightParen,
-	LeftBrace,
-	RightBrace,
-	Semicolon,
-	SingleQuote,
-	Codepoint(char),
-	Identifier(String),
+pub struct TokenList {
+	pub tokens: Vec<lex::Token>,
+	pub context: ast::Context,
 }
 
-#[derive(Debug)]
-pub struct TokenList {
-	pub tokens: Vec<Token>,
+impl TokenList {
+	pub fn new(path: PathBuf, source: String) -> TokenList {
+		Self {
+			tokens: Vec::new(),
+			context: ast::Context::new(path, source)
+		}
+	}
 }
 
 #[derive(Debug)]
 pub struct SyntaxTree {}
 
-pub fn lex(_config: &CompilationConfiguration) -> TokenList {
-	let list: TokenList = TokenList { tokens: Vec::new() };
+pub fn get_source_text(source: ProgramSource) -> (PathBuf, String) {
+	let mut source_text: String = String::new();
+	let mut source_path: PathBuf = PathBuf::new();
+	match source {
+		ProgramSource::Path(pathbuf) => { source_path = pathbuf.clone(); source_text = std::fs::read_to_string(pathbuf.as_path()).unwrap() },
+		ProgramSource::File(mut file) => { source_path = "<file?handle?>".into(); file.read_to_string(&mut source_text); },
+		ProgramSource::Stdin(mut stdin) => { source_path = "<stdin>".into(); stdin.read_to_string(&mut source_text); }
+	};
+	(source_path, source_text)
```

Comment:
Because Rust is expression-based, you can drop your two variables and simply write:

```rs
match source {
  PS::Path(path) => (path.clone(), std::fs::read_to_string(...).unwrap()),
  // Same for the others
}
```

Crashing on file read errors is pretty rude, so maybe this function should return `io::Result<(...)>` instead?